sklearn/utils/validation.py

```diff
@@ -805,6 +805,11 @@ def check_X_y(X, y, accept_sparse=False, *, accept_large_sparse=True,
         y = check_array(y, accept_sparse='csr', force_all_finite=True,
                         ensure_2d=False, dtype=None)
     else:
+        if hasattr(y, 'dtype') and type(y.dtype).__name__ == 'BooleanDtype':
```

Comment:
The `pd.NA` is in all new pandas dtypes. Should we then do the same thing for all of them?

Reply:
Yes, can be done. Should I map each pandas dtype to numpy dtype here?

Reply:
@TomAugspurger do we have a nice mapping on the pandas side to map them to python or numpy dtypes?

Reply:
No, we don't. BooleanArray and IntegerArray provide a `numpy_dtype` that has the NumPy analog.

Reply:
Having the mapping is one way of doing it, another way is to `.to_numpy()` instead and have a `raise_on_na` parameter or something? I know we have talked about it, and I vaguely remember it being fixed? But I can't find it now.

Reply:
> another way is to .to_numpy() instead and have a raise_on_na parameter or something?

What's the desired behavior here? Raise with NAs present, succeed otherwise? I think `np.asarray(dtype=bool)` achieves that.

```python
In [24]: a = pd.array([True, False, pd.NA], dtype="boolean")

In [25]: a
Out[25]:
<BooleanArray>
[True, False, <NA>]
Length: 3, dtype: boolean

In [26]: np.asarray(a[:2], dtype=bool)
Out[26]: array([ True, False])

In [27]: np.asarray(a, dtype=bool)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-27-78a33c9f29bb> in <module>
----> 1 np.asarray(a, dtype=bool)

~/Envs/pandas-dev/lib/python3.7/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
     81
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84
     85

~/sandbox/pandas/pandas/core/arrays/masked.py in __array__(self, dtype)
    217         We return an object array here to preserve our scalar values
    218         """
--> 219         return self.to_numpy(dtype=dtype)
    220
    221     def __arrow_array__(self, type=None):

~/sandbox/pandas/pandas/core/arrays/masked.py in to_numpy(self, dtype, copy, na_value)
    199             ):
    200                 raise ValueError(
--> 201                     f"cannot convert to '{dtype}'-dtype NumPy array "
    202                     "with missing values. Specify an appropriate 'na_value' "
    203                     "for this dtype."

ValueError: cannot convert to 'bool'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```

Reply:
Thanks @TomAugspurger , this is exactly what I was looking for. So we don't really need to check for the dtype. We can just try to convert, and check if this error is raised, or just let the raised `ValueError` to be propagated up to the user.